hudi-spark-datasource/hudi-spark3-extensions_2.12/src/main/scala/org/apache/spark/sql/hudi/execution/CreateHudiTableAsSelectCommand.scala

```diff
@@ -0,0 +1,138 @@
+      }
+      if (mode == SaveMode.Ignore) {
+        // scalastyle:off
+        return Seq.empty
+        // scalastyle:on
+      }
+      // append table
+      saveDataIntoHudiTable(sparkSession, table, path, table.storage.properties, enableHive)
+    } else {
+      val properties = table.storage.properties
+      assert(table.schema.isEmpty)
+      sparkSession.sessionState.catalog.validateTableLocation(table)
+      // create table
+      if (!enableHive) {
+        val newTable = if (!HudiSQLUtils.tableExists(path, conf)) {
+          table.copy(schema = query.schema)
+        } else {
+          table
+        }
+        CreateHudiTableCommand(newTable, true).run(sparkSession)
+      }
+      saveDataIntoHudiTable(sparkSession, table, path, properties, enableHive, "bulk_insert")
+    }
+
+    // save necessary parameter in hoodie.properties
+    val newProperties = new Properties()
+    newProperties.putAll(table.storage.properties.asJava)
+    // add table partition
+    newProperties.put(PARTITIONPATH_FIELD_OPT_KEY, table.partitionColumnNames.mkString(","))
+    val metaPath = new Path(path, HoodieTableMetaClient.METAFOLDER_NAME)
+    val propertyPath = new Path(metaPath, HoodieTableConfig.HOODIE_PROPERTIES_FILE)
+    CreateHudiTableCommand.saveNeededPropertiesIntoHoodie(propertyPath.getFileSystem(conf), propertyPath, newProperties)
+    Seq.empty[Row]
+  }
+
+  private def saveDataIntoHudiTable(
+      session: SparkSession,
+      table: CatalogTable,
+      tablePath: String,
+      tableOptions: Map[String, String],
+      enableHive: Boolean,
+      operationType: String = "upsert"): Unit = {
+    val mode = if (operationType.equals("upsert")) {
+      "append"
+    } else {
+      "overwrite"
+    }
+    val newTableOptions = if (enableHive) {
+      // add hive sync properties
+      val extraOptions = Map("hoodie.table.name" -> table.identifier.table,
+        "path" -> tablePath,
+        DataSourceWriteOptions.HIVE_DATABASE_OPT_KEY -> table.identifier.database.getOrElse("default"),
+        DataSourceWriteOptions.HIVE_TABLE_OPT_KEY -> table.identifier.table,
+        PARTITIONPATH_FIELD_OPT_KEY -> table.partitionColumnNames.mkString(","),
+        DataSourceWriteOptions.HIVE_PARTITION_FIELDS_OPT_KEY -> table.partitionColumnNames.mkString(","),
+        DataSourceWriteOptions.HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY -> "org.apache.hudi.hive.MultiPartKeysValueExtractor")
+      HudiSQLUtils.buildDefaultParameter(extraOptions ++ tableOptions, true)
+    } else {
+      val extraOptions = Map("hoodie.table.name" -> table.identifier.table,
+        "path" -> tablePath,
+        PARTITIONPATH_FIELD_OPT_KEY -> table.partitionColumnNames.mkString(","))
+      HudiSQLUtils.buildDefaultParameter(extraOptions ++ tableOptions, false)
+    }
+
+    HudiSQLUtils.checkWriteOptions(newTableOptions)
+    val df = Dataset.ofRows(session, query)
+    val par = session.sessionState.conf.getConfString("spark.hoodie.shuffle.parallelism", "1500")
```

Comment:
should be `hoodie.bulkinsert.shuffle.parallelism` ?

Reply:
No， this logcial is just a simple way for user to assign parallelism of hudi in sql。
user has no way to assign parallelism of hudi in sql directly。
 if user execute set statement： set spark.hoodie.shuffle.parallelism=4 ， we will grab this value from sessionConf and set it to hudi。
for example：
      spark.sql("set spark.hoodie.shuffle.parallelism=4")
      spark.sql("delete from deleteTable where col1 > 100")
we will set  DELETE_PARALLELISM to 4.